code/datums/components/hattable.dm

```diff
@@ -0,0 +1,110 @@
+TYPEINFO(/datum/component/hattable) // Take a walk through my TWISTED mind.... I'm sorry
+	initialization_args = list(
+		ARG_INFO("death_remove", DATA_INPUT_TEXT, "Remove component and signals on death?", FALSE), // Use this for things that don't revive, or explode on death
+		ARG_INFO("free_hat", DATA_INPUT_TEXT, "Should the parent get a free hat?", FALSE), // For AIs
+	)
+
+/datum/component/hattable  // Compatible stuff should be given hat_offset x and y vars in their own file
+	dupe_mode = COMPONENT_DUPE_HIGHLANDER
+	var/y_level = 0
+	var/x_level = 0
+	var/obj/item/hat = null
+	var/death_throw = FALSE
+	var/death_remove = FALSE
+	var/free_hat = FALSE
+	var/free_hat_lockout = FALSE
+
+/datum/component/hattable/Initialize(death_remove, free_hat)
+	. = ..()
+	RegisterSignal(src.parent, COMSIG_ATTACKBY, PROC_REF(hat_on_obj), override = TRUE)
+	src.death_throw = FALSE
+	src.death_remove = death_remove
+	src.free_hat = free_hat
+	if (free_hat && !free_hat_lockout) // If the thing gets a free hat (currently just AIs with spacebux) and they haven't received it yet, do that
+		var/free_hat_type = pick(childrentypesof(pick(childrentypesof(/obj/item/clothing/head))))
+		hat_on_obj(src.parent, free_hat_type)
+
+
+/datum/component/hattable/proc/hat_on_obj(mob/target as mob, obj/item/item as obj, mob/attacker)
+	var/obj/item/H = null
+	if (src.free_hat && !src.free_hat_lockout)
+		item = new item(get_turf(src)) // Make a new item, lockout future free hats
+		free_hat_lockout = TRUE
+	H = item
+
+	var/atom/movable/hatted = src.parent
+	var/obj/item/hatted_offset = src.parent
+	var/is_item_hat = FALSE
+	if (istype(H, /obj/item/clothing/head/))
+		is_item_hat = TRUE
+		ADD_FLAG(H.appearance_flags, KEEP_TOGETHER) // Flags needed for wigs!
+		ADD_FLAG(H.vis_flags, VIS_INHERIT_DIR)
+
+	if (!src.hat && is_item_hat)
+		if (attacker)
+			attacker.drop_item()
+		else
+			target.drop_item(H) // you're going there whether you like it or not
+
+		src.y_level = 0 // Make ABSOLUTE SURE we're not adding onto a previous offset
+		src.x_level = 0
+		src.y_level = hatted_offset.hat_offset_y + H.hat_offset_y // If a specific hat appears in a strange place, give it its own hat_offset vars to edit the offset by
```

Comment:
see the comment on the AI definition, these should be passed on init, not as vars on specific mob types